packages/swing-store/src/swingStore.js

```diff
@@ -58,30 +71,198 @@ export function makeSnapStoreIO() {
+ *
+ * @property {(includeHistorical: boolean) => AsyncIterable<string>} getArtifactNames
+ *
+ * Get a list of name of artifacts available from the swingStore.  A name returned
+ * by this method guarantees that a call to `getArtifact` on the same exporter
+ * instance will succeed. Options control the filtering of the artifact names
+ * yielded.
+ *
+ * Artifact names:
+ * - transcript.${vatID}.${startPos}.${endPos}
+ * - snapshot.${vatID}.${endPos}
+ *
+ * @property {(name: string) => AsyncIterable<Uint8Array>} getArtifact
+ *
+ * Retrieve an artifact by name.  May throw if the artifact is not available,
+ * which can occur if the artifact is historical and wasn't been preserved.
+ *
+ * @property {() => Promise<void>} close
+ *
+ * Dispose of all resources held by this exporter. Any further operation on this
+ * exporter or its outstanding iterators will fail.
+ */
+
+/**
+ * @param {string} dirPath
+ * @returns {SwingStoreExporter}
+ */
+export function makeSwingStoreExporter(dirPath) {
+  typeof dirPath === 'string' || Fail`dirPath must be a string`;
+  const filePath = path.join(dirPath, 'swingstore.sqlite');
+  const db = sqlite3(filePath);
+
+  // Execute the data export in a (read) transaction, to ensure that we are
+  // capturing the state of the database at a single point in time.
+  const sqlBeginTransaction = db.prepare('BEGIN IMMEDIATE TRANSACTION');
+  sqlBeginTransaction.run();
+
+  const snapStore = makeSnapStore(db, makeSnapStoreIO());
+  const transcriptStore = makeTranscriptStore(
+    db,
+    () => {},
+    () => {},
+  );
+
+  const sqlGetAllKVData = db.prepare(`
+    SELECT key, value
+    FROM kvStore
+    ORDER BY key
+  `);
+
+  /**
+   * @returns {AsyncIterable<KVPair>}
+   * @yields {KVPair}
+   */
+  async function* getExportData() {
+    const kvPairs = sqlGetAllKVData.iterate();
+    for (const kv of kvPairs) {
+      if (getKeyType(kv.key) === 'consensus') {
+        yield [`kv.${kv.key}`, kv.value];
+      }
+    }
+    yield* snapStore.getExportRecords(true);
+    yield* transcriptStore.getExportRecords(true);
+  }
+
+  /**
+   * @param {boolean} includeHistorical
+   * @returns {AsyncIterable<string>}
+   * @yields {string}
+   */
+  async function* getArtifactNames(includeHistorical) {
+    yield* snapStore.getArtifactNames(includeHistorical);
+    yield* transcriptStore.getArtifactNames(includeHistorical);
+  }
+
+  /**
+   * @param {string} name
+   * @returns {AsyncIterable<Uint8Array>}
+   */
+  function getArtifact(name) {
+    typeof name === 'string' || Fail`artifact name must be a string`;
+    const parts = name.split('.');
+    const [type, vatID, pos] = parts;
+
+    if (type === 'snapshot') {
+      // `snapshot.${vatID}.${endPos}`;
+      // prettier-ignore
+      parts.length === 3 ||
+        Fail`expected artifact name of the form 'snapshot.{vatID}.{endPos}', saw ${q(name)}`;
+      return snapStore.exportSnapshot(vatID, Number(pos));
```

Comment:
That happened because `swingStore` already had to split the name string to extract the type, at which point it already had the name components in hand...  I think I agree about the division of responsibility though; easy refactor.